packages/wonder-blocks-data/components/data.js

```diff
@@ -0,0 +1,187 @@
+// @flow
+import * as React from "react";
+import {Server} from "@khanacademy/wonder-blocks-core";
+
+import {RequestFulfillment} from "../util/request-fulfillment.js";
+import {ResponseCache} from "../util/response-cache.js";
+import {TrackerContext} from "../util/request-tracking.js";
+
+import type {Result, IRequestHandler} from "../util/types.js";
+
+type Props<TOptions, TData> = {|
+    handler: IRequestHandler<TOptions, TData>,
+    options: TOptions,
+    children: (result: Result<TData>) => React.Node,
+|};
+
+type State<TData> = {|
+    loading: boolean,
+    data: ?TData,
+    error: ?string,
+|};
+
+export default class Data<TOptions, TData> extends React.Component<
+    Props<TOptions, TData>,
+    State<TData>,
+> {
+    _mounted: boolean;
+
+    constructor(props: Props<TOptions, TData>) {
+        super(props);
+
+        this.state = this._buildStateAndfulfillNeeds();
+    }
+
+    componentDidMount() {
+        this._mounted = true;
+    }
+
+    shouldComponentUpdate(
+        nextProps: $ReadOnly<Props<TOptions, TData>>,
+        nextState: $ReadOnly<State<TData>>,
+    ): boolean {
+        /**
+         * We only bother updating if our state changed.
+         *
+         * And we only update the state if props changed
+         * or we got new data/error.
+         */
+        if (!this._propsMatch(nextProps)) {
+            const newState = this._buildStateAndfulfillNeeds();
+            this.setState(newState);
+        }
+
+        return (
+            this.state.loading !== nextState.loading ||
+            this.state.data !== nextState.data ||
+            this.state.error !== nextState.error
+        );
+    }
+
+    componentWillUnmount() {
+        this._mounted = false;
+    }
+
+    _propsMatch(otherProps: $Shape<Props<TOptions, TData>>): boolean {
+        const {handler, options} = this.props;
+        const {handler: prevHandler, options: prevOptions} = otherProps;
+        return (
+            handler === prevHandler &&
+            handler.getKey(options) === prevHandler.getKey(prevOptions)
+        );
+    }
+
+    _buildStateAndfulfillNeeds(): State<TData> {
+        const propsAtFulfillment = this.props;
+        const {handler, options} = propsAtFulfillment;
+        const {getEntry} = ResponseCache.Default;
+
+        const cachedData = getEntry(handler, options);
+        if (!Server.isServerSide() && cachedData == null) {
+            /**
+             * We're not on the server and the cache missed, or the cache is
+             * invalid.
```

Comment:
If the cache is invalid it feels like the sort of thing we should have logging for - I feel like if we hit this in production it'd be really bad and we'd want to know about it!

Reply:
That isn't something we would track here though. This code has no idea which way that is. That's something the `RequestFulfillment` manages. I'll add something there.

Reply:
I've clarified what this comment means.